File: c/compiler.c
#include "compiler.h"
#include "astNodeOp.h"
#include "tokenType.h"
#include "astNode.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>
#include <llvm-c/Types.h>
#include <llvm-c/Core.h>
#include <llvm-c/Target.h>
#include <llvm-c/TargetMachine.h>
#include <llvm-c/Analysis.h>

const int TEXT_LEN_LIMIT = 512;

Compiler Compiler_new(char* filename) {
    Compiler comp;
    comp.inFile = fopen(filename, "r");

    if (!comp.inFile) {
        fprintf(stderr, "Unable to load file %s\n", filename);
        exit(1);
    }

    comp.line = 1;
    comp.putback = '\n';
    comp.token.type = T_EOF;
    comp.token.intValue = 0;
    comp.globals = g_hash_table_new(g_str_hash, g_str_equal);
    return comp;
}

char Compiler_next(Compiler* comp) {
    char c;

    if (comp->putback) {
        c = comp->putback;
        comp->putback = '\0';
        return c;
    }

    c = fgetc(comp->inFile);

    if (c == '\n') {
        comp->line++;
    } else if (c == EOF) {
        c = '\0';
    }

    return c;
}

char Compiler_skip(Compiler* comp) {
    char c = Compiler_next(comp);

    while ((c == ' ') || (c == '\t') || (c == '\n') || (c == '\r') || (c == '\f')) {
        c = Compiler_next(comp);
    }

    return c;
}

bool Compiler_scan(Compiler* comp) {
    char c = Compiler_skip(comp);

    switch (c) {
        case '\0':
            comp->token.type = T_EOF;
            return false;
        case '+':
            comp->token.type = T_PLUS;
            break;
        case '-':
            comp->token.type = T_MINUS;
            break;
        case '*':
            comp->token.type = T_STAR;
            break;
        case '/':
            comp->token.type = T_SLASH;
            break;
        case ';':
            comp->token.type = T_SEMI;
            break;
        case '=':
            if ((c = Compiler_next(comp)) == '=') {
                comp->token.type = T_EQUAL;
            } else {
                comp->putback = c;
                comp->token.type = T_ASSIGN;
            }

            break;
        case '!':
            if ((c = Compiler_next(comp)) == '=') {
                comp->token.type = T_NOTEQUAL;
            } else {
                fprintf(stderr, "Unrecognized character:%c on line %d\n", c, comp->line);
                exit(1);
            }

            break;
        case '<':
            if ((c = Compiler_next(comp)) == '=') {
                comp->token.type = T_LESSEQUAL;
            } else {
                comp->putback = c;
                comp->token.type = T_LESSTHAN;
            }

            break;
        case '>':
            if ((c = Compiler_next(comp)) == '=') {
                comp->token.type = T_GREATEREQUAL;
            } else {
                comp->putback = c;
                comp->token.type = T_GREATERTHAN;
            }

            break;
        default:
            if (isdigit(c)) {
                comp->token.intValue = Compiler_scanint(comp, c);
                comp->token.type = T_INTLIT;
                break;
            } else if ((isalpha(c)) || (c == '_')) {
                Compiler_scanident(comp, c, comp->text);
                TokenType newTokenType = Compiler_keyword(comp, comp->text);

                if (newTokenType != T_EOF) {
                    comp->token.type = newTokenType;
                    break;
                } else {
                    comp->token.type = T_IDENT;
                    break;
                }
            }

            fprintf(stderr, "Unrecognized character %c on line %d\n", c, comp->line);
            exit(1);
    }

    return true;
}

int chrpos(char* s, char c) {
    if (c == '\0') {
        return -1;
    }

    char* p = strchr(s, c);
    return (p ? p - s : -1);
}

int Compiler_scanint(Compiler* comp, char c) {
    int k, val = 0;

    while ((k = chrpos("0123456789", c)) >= 0) {
        val = val * 10 + k;
        c = Compiler_next(comp);
    }

    comp->putback = c;
    return val;
}

void Compiler_scanident(Compiler* comp, char c, char* buf) {
    int i = 0;

    while ((isalpha(c)) || (isdigit(c)) || (c == '_')) {
        if (i == (TEXT_LEN_LIMIT - 1)) {
            fprintf(stderr, "identifier too long on line %d\n", comp->line);
            exit(1);
        } else if (i < (TEXT_LEN_LIMIT - 1)) {
            buf[i++] = c;
        }

        c = Compiler_next(comp);
    }

    comp->putback = c;
    buf[i] = '\0';
}

TokenType Compiler_keyword(Compiler* comp, char* s) {
    switch (s[0]) {
        case 'p':
            if (strcmp(s, "print\0") == 0) {
                return T_PRINT;
            }

            break;
        case 'i':
            if (strcmp(s, "int\0") == 0) {
                return T_INT;
            }

            break;
    }

    return T_EOF;
}

void Compiler_match(Compiler* comp, TokenType ttype, char* tstr) {
    if (comp->token.type == ttype) {
        Compiler_scan(comp);
    } else {
        fprintf(stderr, "%s expected on line %d\n", tstr, comp->line);
        exit(1);
    }
}

void Compiler_semi(Compiler* comp) {
    Compiler_match(comp, T_SEMI, ";");
}

void Compiler_ident(Compiler* comp) {
    Compiler_match(comp, T_IDENT, "identifier");
}

int opPrec[] = {
    0, 10, 10,      // T_EOF, T_PLUS, T_MINUS
    20, 20,         // T_STAR, T_SLASH
    30, 30,         // T_EQUAL, T_NOTEQUAL
    40, 40, 40, 40  // T_LESSTHAN, T_GREATERTHAN, T_LESSEQUAL, T_GREATEREQUAL
};

int Compiler_op_precedence(Compiler* comp, TokenType type) {
    int prec = opPrec[type];

    if (prec == 0) {
        fprintf(stderr, "Syntax error, token:%d on line %d\n", type, comp->line);
        exit(1);
    }

    return prec;
}

ASTNodeOp Compiler_arithop(Compiler* comp, TokenType tok) {
    if ((tok > T_EOF) && (tok < T_INTLIT)) {
        return (ASTNodeOp)tok;
    }

    fprintf(stderr, "Syntax error, token:%d on line %d\n", tok, comp->line);
    exit(1);
}

ASTNode* Compiler_primary(Compiler* comp) {
    ASTNode* node;
    LLVMValueRef id;
    
    switch (comp->token.type) {
        case T_INTLIT:
            node = mkAstLeaf_int(A_INTLIT, comp->token.intValue);
            break;
        case T_IDENT:
            id = Compiler_find_global(comp, comp->text);

            if (id == NULL) {
                fprintf(stderr, "Unknown variable:%s on line %d\n", comp->text, comp->line);
                exit(1);
            }

            node = mkAstLeaf_value(A_IDENT, id);
            break;
        default:
            fprintf(stderr, "Syntax error on line %d\n", comp->line);
            exit(1);
    }

    
    Compiler_scan(comp);
    return node;
}

ASTNode* Compiler_binexpr(Compiler* comp, int ptp) {
    ASTNode* left = Compiler_primary(comp);
    TokenType tokenType = comp->token.type;

    if (tokenType == T_SEMI) {
        return left;
    }

    while (Compiler_op_precedence(comp, tokenType) > ptp) {
        Compiler_scan(comp);
        ASTNode* right = Compiler_binexpr(comp, opPrec[tokenType]);
        left = mkAstNode_int(Compiler_arithop(comp, tokenType), left, right, 0);
        tokenType = comp->token.type;

        if (tokenType == T_SEMI) {
            return left;
        }
    }

    return left;
}

void Compiler_statements(Compiler* comp, LLVMBuilderRef builder, LLVMContextRef context, LLVMTypeRef printf_type, LLVMValueRef printf_func) {
    while (true) {
        switch (comp->token.type) {
            case T_PRINT:
                Compiler_print_statement(comp, builder, context, printf_type, printf_func);
                break;
            case T_INT:
                Compiler_var_declaration(comp, builder, context);
                break;
            case T_IDENT:
                Compiler_assignment_statement(comp, builder, context);
                break;
            case T_EOF:
                return;
            default:
                fprintf(stderr, "Syntax error, token:%d on line %d\n", comp->token.type, comp->line);
                exit(1);
        }
    }
}

void Compiler_print_statement(Compiler* comp, LLVMBuilderRef builder, LLVMContextRef context, LLVMTypeRef printf_type, LLVMValueRef printf_func) {
    Compiler_match(comp, T_PRINT, "print");
    ASTNode* node = Compiler_binexpr(comp, 0);
    LLVMValueRef ret_val = Compiler_buildAST(comp, node, builder, context);
    Compiler_generatePrint(comp, builder, context, printf_type, printf_func, ret_val);
    Compiler_semi(comp);
}

void Compiler_var_declaration(Compiler* comp, LLVMBuilderRef builder, LLVMContextRef context) {
    Compiler_match(comp, T_INT, "int");
    Compiler_ident(comp);
    Compiler_add_global(comp, comp->text, builder, context);
    Compiler_semi(comp);
}

void Compiler_assignment_statement(Compiler* comp, LLVMBuilderRef builder, LLVMContextRef context) {
    Compiler_ident(comp);
    LLVMValueRef id;
    
    if ((id = Compiler_find_global(comp, comp->text)) == NULL) {
        fprintf("Undeclared variable:%s on line %d\n", comp->text, comp->line);
        exit(1);
    }

    ASTNode* right = mkAstLeaf_value(A_LVIDENT, id);
    Compiler_match(comp, T_ASSIGN, "=");
    ASTNode* left = Compiler_binexpr(comp, 0);
    ASTNode* tree = mkAstNode_int(A_ASSIGN, left, right, 0);
    Compiler_buildAST(comp, tree, builder, context);
    Compiler_semi(comp);
}

void Compiler_add_global(Compiler* comp, char* global_var, LLVMBuilderRef builder, LLVMContextRef context) {
    LLVMValueRef val = LLVMBuildAlloca(builder, LLVMInt32TypeInContext(context), global_var);
    g_hash_table_insert(comp->globals, global_var, val);
}

LLVMValueRef Compiler_find_global(Compiler* comp, char* global_var) {
    return g_hash_table_lookup(comp->globals, global_var);
}

LLVMValueRef Compiler_buildAST(Compiler* comp, ASTNode* node, LLVMBuilderRef builder, LLVMContextRef context) {
    LLVMValueRef leftVal;
    LLVMValueRef rightVal;
    LLVMValueRef icmpVal;

    if (node->left) {
        leftVal = Compiler_buildAST(comp, node->left, builder, context);
        free(node->left);
    }

    if (node->right) {
        rightVal = Compiler_buildAST(comp, node->right, builder, context);
        free(node->right);
    }

    switch (node->op) {
        case A_ADD:
            return LLVMBuildAdd(builder, leftVal, rightVal, "add");
        case A_SUBTRACT:
            return LLVMBuildSub(builder, leftVal, rightVal, "sub");
        case A_MULTIPLY:
            return LLVMBuildMul(builder, leftVal, rightVal, "mul");
        case A_DIVIDE:
            return LLVMBuildSDiv(builder, leftVal, rightVal, "div");
        case A_INTLIT:
            return LLVMConstInt(LLVMInt32TypeInContext(context), node->intValue, false);
        case A_LVIDENT:
            if (node->nodeType != ANT_VALUE) {
                fprintf(stderr, "Invalid node type:%d\n", node->nodeType);
                exit(1);
            }

            return node->val;
        case A_ASSIGN:
            LLVMBuildStore(builder, leftVal, rightVal);
            return NULL;
        case A_IDENT:
            return LLVMBuildLoad2(builder, LLVMInt32TypeInContext(context), node->val, "load");
        case A_EQUAL:
            icmpVal = LLVMBuildICmp(builder, LLVMIntEQ, leftVal, rightVal, "icmp");
            return LLVMBuildZExt(builder, icmpVal, LLVMInt32TypeInContext(context), "zext");
        case A_NOTEQUAL:
            icmpVal = LLVMBuildICmp(builder, LLVMIntNE, leftVal, rightVal, "icmp");
            return LLVMBuildZExt(builder, icmpVal, LLVMInt32TypeInContext(context), "zext");
        case A_LESSTHAN:
            icmpVal = LLVMBuildICmp(builder, LLVMIntSLT, leftVal, rightVal, "icmp");
            return LLVMBuildZExt(builder, icmpVal, LLVMInt32TypeInContext(context), "zext");
        case A_GREATERTHAN:
            icmpVal = LLVMBuildICmp(builder, LLVMIntSGT, leftVal, rightVal, "icmp");
            return LLVMBuildZExt(builder, icmpVal, LLVMInt32TypeInContext(context), "zext");
        case A_LESSEQUAL:
            icmpVal = LLVMBuildICmp(builder, LLVMIntSLE, leftVal, rightVal, "icmp");
            return LLVMBuildZExt(builder, icmpVal, LLVMInt32TypeInContext(context), "zext");
        case A_GREATEREQUAL:
            icmpVal = LLVMBuildICmp(builder, LLVMIntSGE, leftVal, rightVal, "icmp");
            return LLVMBuildZExt(builder, icmpVal, LLVMInt32TypeInContext(context), "zext");
        default:
            fprintf(stderr, "unrecognized node in ast %d\n", node->op);
            exit(1);
    }
}

void Compiler_generatePrint(Compiler* comp, LLVMBuilderRef builder, LLVMContextRef context, LLVMTypeRef printf_type, LLVMValueRef printf_func, LLVMValueRef value) {
    LLVMValueRef format_const = LLVMConstStringInContext(context, "%d\n", 4, true);
    LLVMValueRef var_ptr = LLVMBuildAlloca(builder, LLVMArrayType(LLVMInt8TypeInContext(context), 4), "fmt_ptr");
    LLVMBuildStore(builder, format_const, var_ptr);
    LLVMValueRef fmt_arg = LLVMBuildBitCast(builder, var_ptr, LLVMPointerType(LLVMInt8TypeInContext(context), 0), "fmt_ptr_bitcast");
    LLVMValueRef printf_args[] = {fmt_arg, value};
    LLVMBuildCall2(builder, printf_type, printf_func, printf_args, 2, "printf_res");
}

void Compiler_parse(Compiler* comp) {
    // Initialize everything
    LLVMInitializeAllTargetInfos();
    LLVMInitializeAllTargets();
    LLVMInitializeAllTargetMCs();
    LLVMInitializeAllAsmParsers();
    LLVMInitializeAllAsmPrinters();

    // Create context
    LLVMContextRef main_context = LLVMContextCreate();

    // Create module
    LLVMModuleRef main_module = LLVMModuleCreateWithNameInContext("main_module", main_context);

    // Set module triple
    char* triple = LLVMGetDefaultTargetTriple();
    LLVMSetTarget(main_module, triple);

    // Create target from triple
    LLVMTargetRef target;
    char* errorMessage;
    LLVMBool failure = LLVMGetTargetFromTriple(triple, &target, &errorMessage);

    if (failure) {
        fprintf(stderr, "Error getting host target: %s\n", errorMessage);
        LLVMDisposeMessage(errorMessage);
        LLVMDisposeModule(main_module);
        LLVMContextDispose(main_context);
        exit(1);
    }

    // Create target machine
    char* cpu_name = LLVMGetHostCPUName();
    char* cpu_features = LLVMGetHostCPUFeatures();
    LLVMTargetMachineRef machine = LLVMCreateTargetMachine(target, triple, cpu_name, cpu_features, LLVMCodeGenLevelNone, LLVMRelocDefault, LLVMCodeModelDefault);
    LLVMSetModuleDataLayout(main_module, LLVMCreateTargetDataLayout(machine));

    // Create builder
    LLVMBuilderRef builder = LLVMCreateBuilderInContext(main_context);

    // Setup printf function
    LLVMTypeRef printf_arg_types[] = {LLVMPointerType(LLVMInt8TypeInContext(main_context), 0)};
    LLVMTypeRef printf_type = LLVMFunctionType(LLVMInt32TypeInContext(main_context), printf_arg_types, 1, true);
    LLVMValueRef printf_func = LLVMAddFunction(main_module, "printf", printf_type);
    LLVMSetLinkage(printf_func, LLVMExternalLinkage);

    // Setup main function
    LLVMTypeRef main_arg_types[] = {LLVMInt32TypeInContext(main_context), LLVMPointerType(LLVMPointerType(LLVMInt8TypeInContext(main_context), 0), 0)};
    LLVMTypeRef main_type = LLVMFunctionType(LLVMInt32TypeInContext(main_context), main_arg_types, 2, false);
    LLVMValueRef main_func = LLVMAddFunction(main_module, "main", main_type);
    LLVMBasicBlockRef main_bb = LLVMAppendBasicBlockInContext(main_context, main_func, "entry");
    LLVMPositionBuilderAtEnd(builder, main_bb);

    // Compile code
    Compiler_statements(comp, builder, main_context, printf_type, printf_func);

    // Return result from main
    LLVMBuildRet(builder, LLVMConstInt(LLVMInt32TypeInContext(main_context), 0, false));

    // Make sure the function is fine
    LLVMVerifyFunction(main_func, LLVMAbortProcessAction);

    // Write to file
    failure = LLVMTargetMachineEmitToFile(machine, main_module, "output.o", LLVMObjectFile, &errorMessage);

    if (failure) {
        fprintf(stderr, "Error writing code to output.o: %s\n", errorMessage);
        LLVMDisposeMessage(errorMessage);
        LLVMDisposeModule(main_module);
        LLVMContextDispose(main_context);
        exit(1);
    }

    LLVMDisposeBuilder(builder);
    LLVMDisposeTargetMachine(machine);
    LLVMDisposeModule(main_module);
    LLVMContextDispose(main_context);
}

void Compiler_run(Compiler* comp) {
    Compiler_scan(comp);
    Compiler_parse(comp);
    fclose(comp->inFile);
    g_hash_table_destroy(comp->globals);
}

File: julia/src/compiler.jl
const TEXT_LEN_LIMIT = 512

mutable struct Compiler
    line::Int64
    putback::Char
    inFile::IOStream
    token::Token
    text::String
    globals::Dict{String, LLVM.Value}

    function Compiler(i::IOStream)
        new(1, '\n', i, Token(T_EOF, 0), "", Dict())
    end
end

function compiler_next(comp::Compiler)
    c = '\0'

    if comp.putback != '\0'
        c = comp.putback
        comp.putback = '\0'
        return c
    end

    try
        c = read(comp.inFile, Char)
    catch err
        if isa(err, EOFError)
            c = '\0'
        else
            throw(err)
        end
    end

    if c == '\n'
        comp.line += 1
    end

    return c
end

function compiler_skip(comp::Compiler)
    c = compiler_next(comp)

    while (c == ' ') || (c == '\t') || (c == '\n') || (c == '\r') || (c == '\f')
        c = compiler_next(comp)
    end

    return c
end

function compiler_scan(comp::Compiler)
    c = compiler_skip(comp)

    if c == '\0'
        comp.token.type = T_EOF
        return false
    elseif c == '+'
        comp.token.type = T_PLUS
    elseif c == '-'
        comp.token.type = T_MINUS
    elseif c == '*'
        comp.token.type = T_STAR
    elseif c == '/'
        comp.token.type = T_SLASH
    elseif c == ';'
        comp.token.type = T_SEMI
    elseif c == '='
        c = compiler_next(comp)

        if c == '='
            comp.token.type = T_EQUAL
        else
            comp.putback = c
            comp.token.type = T_ASSIGN
        end
    elseif c == '!'
        c = compiler_next(comp)

        if c == '='
            comp.token.type = T_NOTEQUAL
        else
            throw(ErrorException("Unrecognized character:" * c * " on line " * string(comp.line)))
        end
    elseif c == '<'
        c = compiler_next(comp)

        if c == '='
            comp.token.type = T_LESSEQUAL
        else
            comp.putback = c
            comp.token.type = T_LESSTHAN
        end
    elseif c == '>'
        c = compiler_next(comp)

        if c == '='
            comp.token.type = T_GREATEREQUAL
        else
            comp.putback = c
            comp.token.type = T_GREATERTHAN
        end
    else
        if isdigit(c)
            comp.token.intValue = compiler_scanint(comp, c)
            comp.token.type = T_INTLIT
            return true
        elseif (isletter(c)) || (c == '_')
            comp.text = compiler_scanident(comp, c)
            tokenType = compiler_keyword(comp.text)
            
            if tokenType != nothing
                comp.token.type = tokenType
            else
                comp.token.type = T_IDENT
            end
        else
            throw(ErrorException("Unrecognized character " * c * " on line " * comp.line))
        end
    end

    return true
end

function chrpos(s::String, c::Char)
    res = findfirst(isequal(c), s)

    if res == nothing
        return -1
    end

    return res - 1
end

function compiler_scanint(comp::Compiler, c::Char)
    k = chrpos("0123456789", c)
    val = 0

    while k >= 0
        val = val * 10 + k
        c = compiler_next(comp)
        k = chrpos("0123456789", c)
    end

    comp.putback = c
    return val
end

function compiler_scanident(comp::Compiler, c::Char)
    i = 0
    buf = ""

    while (isletter(c)) || (isdigit(c)) || (c == '_')
        if i == (TEXT_LEN_LIMIT - 1)
            throw(ErrorException("identifier too long on line " * string(comp.line)))
        elseif i < (TEXT_LEN_LIMIT - 1)
            buf *= c
        end

        c = compiler_next(comp)
    end

    comp.putback = c
    buf *= '\0'
end

function compiler_keyword(s::String)
    if s == "print\0"
        return T_PRINT
    elseif s == "int\0"
        return T_INT
    end

    return nothing
end

function compiler_arithop(comp::Compiler, token::TokenType)
    if token == T_PLUS
        return A_ADD
    elseif token == T_MINUS
        return A_SUBTRACT
    elseif token == T_STAR
        return A_MULTIPLY
    elseif token == T_SLASH
        return A_DIVIDE
    elseif token == T_EQUAL
        return A_EQUAL
    elseif token == T_NOTEQUAL
        return A_NOTEQUAL
    elseif token == T_LESSTHAN
        return A_LESSTHAN
    elseif token == T_GREATERTHAN
        return A_GREATERTHAN
    elseif token == T_LESSEQUAL
        return A_LESSEQUAL
    elseif token == T_GREATEREQUAL
        return A_GREATEREQUAL
    else
        throw(ErrorException("Syntax error, token:" * string(token) * " on line " * string(comp.line)))
    end
end

function compiler_primary(comp::Compiler)
    if comp.token.type == T_INTLIT
        node = ASTNode(A_INTLIT, comp.token.intValue)
        compiler_scan(comp)
        return node
    elseif comp.token.type == T_IDENT
        id = compiler_find_global(comp, comp.text)

        if id == nothing
            throw(ErrorException("Unknown variable:" * comp.text * " on line " * string(comp.line)))
        end

        node = ASTNode(A_IDENT, id)
        compiler_scan(comp)
        return node
    else
        throw(ErrorException("Syntax error on line " * string(comp.line) * " - token " * string(comp.token.type)))
    end
end

function compiler_opPrec(comp::Compiler, tokenType::TokenType)
    if tokenType == T_EOF
        throw(ErrorException("Syntax error, token:" * string(tokenType) * " on line " * string(comp.line)))
    elseif (tokenType == T_PLUS) || (tokenType == T_MINUS)
        return 10
    elseif (tokenType == T_STAR) || (tokenType == T_SLASH)
        return 20
    elseif (tokenType == T_EQUAL) || (tokenType == T_NOTEQUAL)
        return 30
    elseif (tokenType == T_LESSTHAN) || (tokenType == T_GREATERTHAN) || (tokenType == T_LESSEQUAL) || (tokenType == T_GREATEREQUAL)
        return 40
    end
end

function compiler_binexpr(comp::Compiler, ptp::Int64)
    left = compiler_primary(comp)
    tokenType = comp.token.type

    if tokenType == T_SEMI
        return left
    end

    op_precedence = compiler_opPrec(comp, tokenType)

    while compiler_opPrec(comp, tokenType) > ptp
        compiler_scan(comp)
        right = compiler_binexpr(comp, op_precedence)
        left = ASTNode(compiler_arithop(comp, tokenType), left, right, 0)
        tokenType = comp.token.type

        if tokenType == T_SEMI
            return left
        end

        op_precedence = compiler_opPrec(comp, tokenType)
    end

    return left
end

function compiler_match(comp::Compiler, type::TokenType, str::String)
    if comp.token.type == type
        compiler_scan(comp)
    else
        throw(ErrorException(str * " expected on line " * string(comp.line)))
    end
end

function compiler_semi(comp::Compiler)
    compiler_match(comp, T_SEMI, ";")
end

function compiler_ident(comp::Compiler)
    compiler_match(comp, T_IDENT, "identifier")
end

function compiler_statements(comp::Compiler, builder::LLVM.Builder, ctx::LLVM.Context, printf_func::LLVM.Function)
    while true
        if comp.token.type == T_PRINT
            compiler_print_statement(comp, builder, ctx, printf_func)
        elseif comp.token.type == T_INT
            compiler_var_declaration(comp, builder, ctx)
        elseif comp.token.type == T_IDENT
            compiler_assignment_statement(comp, builder, ctx)
        elseif comp.token.type == T_EOF
            return
        else
            throw(ErrorException("Syntax error, token:" * string(comp.token.type) * " on line " * string(comp.line)))
        end
    end
end

function compiler_print_statement(comp::Compiler, builder::LLVM.Builder, ctx::LLVM.Context, printf_func::LLVM.Function)
    compiler_match(comp, T_PRINT, "print")
    tree = compiler_binexpr(comp, 0)
    ret_val = compiler_buildAST(comp, tree, builder, ctx)
    compiler_print(comp, ret_val, ctx, builder, printf_func)
    compiler_semi(comp)
end

function compiler_var_declaration(comp::Compiler, builder::LLVM.Builder, ctx::LLVM.Context)
    compiler_match(comp, T_INT, "int")
    compiler_ident(comp)
    compiler_add_global(comp, comp.text, builder, ctx)
    compiler_semi(comp)
end

function compiler_assignment_statement(comp::Compiler, builder::LLVM.Builder, ctx::LLVM.Context)
    compiler_ident(comp)
    id = compiler_find_global(comp, comp.text)

    if id == nothing
        throw(ErrorException("Undeclared variable:" * comp.text * " on line " * string(comp.line)))
    end

    right = ASTNode(A_LVIDENT, id)
    compiler_match(comp, T_ASSIGN, "=")
    left = compiler_binexpr(comp, 0)
    tree = ASTNode(A_ASSIGN, left, right, 0)
    compiler_buildAST(comp, tree, builder, ctx)
    compiler_semi(comp)
end

function compiler_add_global(comp::Compiler, global_var::String, builder::LLVM.Builder, ctx::LLVM.Context)
    comp.globals[global_var] = alloca!(builder, LLVM.Int32Type(ctx))
end

function compiler_find_global(comp::Compiler, global_var::String)
    return get(comp.globals, global_var, nothing)
end

function compiler_buildAST(comp::Compiler, node::ASTNode, builder::LLVM.Builder, ctx::LLVM.Context)
    leftValue = nothing
    rightValue = nothing

    if node.left != nothing
        leftValue = compiler_buildAST(comp, node.left, builder, ctx)
    end

    if node.right != nothing
        rightValue = compiler_buildAST(comp, node.right, builder, ctx)
    end

    if node.op == A_ADD
        return add!(builder, leftValue, rightValue)
    elseif node.op == A_SUBTRACT
        return sub!(builder, leftValue, rightValue)
    elseif node.op == A_MULTIPLY
        return mul!(builder, leftValue, rightValue)
    elseif node.op == A_DIVIDE
        return sdiv!(builder, leftValue, rightValue)
    elseif node.op == A_INTLIT
        val = alloca!(builder, LLVM.Int32Type(ctx))
        store!(builder, ConstantInt(LLVM.Int32Type(ctx), node.value), val)
        return load!(builder, val)
    elseif node.op == A_LVIDENT
        return node.value
    elseif node.op == A_ASSIGN
        store!(builder, leftValue, rightValue)
    elseif node.op == A_IDENT
        return load!(builder, node.value)
    elseif node.op == A_EQUAL
        val = icmp!(builder, LLVM.API.LLVMIntEQ, leftValue, rightValue)
        return zext!(builder, val, LLVM.Int32Type(ctx))
    elseif node.op == A_NOTEQUAL
        val = icmp!(builder, LLVM.API.LLVMIntNE, leftValue, rightValue)
        return zext!(builder, val, LLVM.Int32Type(ctx))
    elseif node.op == A_LESSTHAN
        val = icmp!(builder, LLVM.API.LLVMIntSLT, leftValue, rightValue)
        return zext!(builder, val, LLVM.Int32Type(ctx))
    elseif node.op == A_LESSEQUAL
        val = icmp!(builder, LLVM.API.LLVMIntSLE, leftValue, rightValue)
        return zext!(builder, val, LLVM.Int32Type(ctx))
    elseif node.op == A_GREATERTHAN
        val = icmp!(builder, LLVM.API.LLVMIntSGT, leftValue, rightValue)
        return zext!(builder, val, LLVM.Int32Type(ctx))
    elseif node.op == A_GREATEREQUAL
        val = icmp!(builder, LLVM.API.LLVMIntSGE, leftValue, rightValue)
        return zext!(builder, val, LLVM.Int32Type(ctx))
    else
        throw(ErrorException("Invalid op in buildAST " * string(node.op)))
    end
end

function compiler_print(comp::Compiler, val::LLVM.Value, ctx::LLVM.Context, builder::LLVM.Builder, printf_func::LLVM.Function)
    format_string = "%d\n\0"
    format_string_val = LLVM.ConstantArray(encode(format_string, "utf-8"); ctx)
    format_string_ptr = alloca!(builder, LLVM.ArrayType(LLVM.Int8Type(ctx), length(format_string)))
    format_string_ptr_ref = LLVM.Value(format_string_ptr.ref)
    store!(builder, format_string_val, format_string_ptr_ref)
    format_arg = bitcast!(builder, format_string_ptr_ref, LLVM.PointerType(LLVM.Int8Type(ctx)))
    call!(builder, printf_func, [format_arg, val])
end

function compiler_parse(comp::Compiler)
    @dispose ctx=Context() begin
        mod = LLVM.Module("main_module"; ctx)

        printf_param_types = [LLVM.PointerType(LLVM.Int8Type(ctx))]
        printf_ret_type = LLVM.Int32Type(ctx)
        printf_type = LLVM.FunctionType(printf_ret_type, printf_param_types; vararg=true)
        printf_func = LLVM.Function(mod, "printf", printf_type)
        linkage!(printf_func, LLVM.API.LLVMExternalLinkage)

        main_param_types = [LLVM.Int32Type(ctx), LLVM.PointerType(LLVM.PointerType(LLVM.Int8Type(ctx)))]
        main_ret_type = LLVM.Int32Type(ctx)
        main_type = LLVM.FunctionType(main_ret_type, main_param_types)
        main_func = LLVM.Function(mod, "main", main_type)

        @dispose builder=Builder(ctx) begin
            main_bb = BasicBlock(main_func, "entry"; ctx)
            position!(builder, main_bb)
            compiler_statements(comp, builder, ctx, printf_func)
            retval = alloca!(builder, LLVM.Int32Type(ctx))
            retval_ref = LLVM.Value(retval.ref)
            store!(builder, ConstantInt(LLVM.Int32Type(ctx), 0), retval_ref)
            retint = load!(builder, retval_ref)
            ret!(builder, retint)
            verify(mod)
        end

        host_triple = Sys.MACHINE
        target = LLVM.Target(triple=host_triple)
        @dispose machine=LLVM.TargetMachine(target, host_triple) begin
            LLVM.emit(machine, mod, LLVM.API.LLVMObjectFile, "output.o")
        end
    end
end